### ISSUE TITLE
Observing Array of References - When Referenced Object Changes, No Rerender?

### ISSUE THREAD

**Issue by mashaalmemon** [xaviergonz/mobx-keystone#390]
I have a situation where I have an array of references to models. When a model being referred to is updated, the observer doesn't seem to rerender.

I've got a simple sample below. Please note this isn't a working example. I've just giving this example to illustrate the idea of what I am trying to achieve, and what is not working. 

mobx-keystone model for an `Item`
```typescript
class Item extends Model {
  myVal:prop<string>("").withSetter()
}
{}
```

 mobx-keystone `ItemList` model which is a list of references to `Item` models
```typescript
class ItemList extends Model {
  itemListRefs:prop<Ref<Item>[]>(()=>[])
}
{

  /** 
   *Setting an item list involves adding references to list items.
   * In this trivial example it doesn't make sense but imagine in my case
   * that we have multiple lists referencing the same items, and so keeping
   */ a list of references makes the most sense.
  @modelAction
  public setItemList(items:Item[]) {
    let this.itemListRefs[] = [];
    for (item of ItemListRefs) {
      let itemRef:Ref<Item> = // make new reference 
      this.itemListRefs.push(itemRef);
    }
  }

  /** 
   * This computed value returns a list of actual models but dereferencing and creating an array of actual
   * Item models.
   */ 
  @computed
  public items():Item[] {
   let items:Item[] = []; 
   for (ref of this.itemListRefs) {
     // simple dereferencing reference and pushing model onto the stack 
     items.push(ref.current);
   }
   return items;
  }
}
```

`ItemListing` react component which observes the `@computed itemList` method. It is an observer but it is observing the list specifically, not necessarily the `Item` in the list. 

```tsx
export const ItemListing: React.FC<any> = observer({
}: any): JSX.Element => {

  // assume I am grabbing items from my root store via prescribed approach
  const {itemList} = RootStore;

  //Render the UI
  return (
    <>
       () =>
      itemList.items.map(
     
        (item, key: number): React.ReactNode => (
          <ItemRow item={item}/>
        )
      )
    </>
  );
});
```

 `ItemRow` react component which takes in an `Item` model as a property. It is not an observer as its getting a property passed in.

```tsx
export const ItemRow: React.FC<any> ={
item:Item
}: any): JSX.Element => {

  
  return (
    <>
       My Value is ${item.myVal}
    </>
  );
};
```

Note that the `ItemListing` component has an observer. But the `Items` in the list around observed directly. 

**What is working:**
* Initial render is fine.
* As items are added to the `ItemList` model, the number of rows are changing as expected. 

**What is not working:**
* However, when the `myVal` prop of an `Item` model that is already rendered and on the screen, is updated, a rerender with the new value being triggered. 

**What I have checked/confirmed:**
* If I view the contents of the model it has the updated value, so have confirmed that underneath the hood all is working as it should.

Any guidance you can give me on the appropriate approach would be ideal.

**Comment by xaviergonz** [comment]
does ItemRow read any properties from inside the item? if so you will either need to use observer on that component too or pass getSnapshot(item) as property

**Comment by mashaalmemon** [comment]
Thanks for the quick response @xaviergonz .

`ItemRow` does read properties from the item object. I had a typo which I fixed.

I tried putting an `observer` on the `ItemRow` component but it didn't seem to react to changes. I tried `getSnapshot(item)` but it also didn't seem to make a difference.

Interestingly, I tried this:
* I made a change to `myVal` on an `Item` Model using a setter. I read back the`myVal` from the `Item` model and it was as expected.
* I then, on the next line,  tried `getSnapshot(item)` on the same model instance and took a look at `myVal` but it had the previous value in it.

**Comment by mashaalmemon** [comment]
Did a little bit more poking around. Putting the `observer` on `ItemRow` seems to have done the trick **partially**. I say partially because on the first render of an `ItemRow` changes to the `Item` model are not being rendered. But after the **second** render of the `ItemRow` suddenly changes are being applied.

Let me elaborate:
* In my case, I have a 'search' for the listing of items. So depending on what I search for, an `ItemRow` may appear in the `ItemListing`. If I adjust the search the `ItemRow` may disappear/appear, and so on. 
* I've found that on my first render of an `ItemRow` which is observing an `Item` model, any changes that are made to the model do not get immediately rendered (although they are applied to the internal state). Only the initial render information is present.
* But if I adjust the search to remove the `ItemRow` and then bring back the same row on a second render (i.e. type "user", then delete the text, then type "user" again, to allow the same `Item` and thus identical `ItemRow` to be rendered from the same normalized model), now all changes made to the model are rendering immediately. 

I've looked at the objects being rendered in the `ItemRow` but haven't immediately seen a difference between the model on first/second render. 

I've never experienced this before with plain Mobx. I can't understand what could possibly be preventing the observer from getting notified on that first render.

**Comment by mashaalmemon** [comment]
Essentially we're seeing the underlying snapshot changing, but despite `observer` being used on the component, it simply isn't reacting and rerendering.

However if add a `useEffect` dependency on any part of the model explicitly, including the `$`/snapshot or `getSnapshot`, a re-render happens.  Should a `useEffect` on the `Item` model (not model `myVal`) should trigger a re-render correct? 

Unfortunately using `getSnapshot` doesn't work well for our usecase because we have custom transforms setup (`Moment` <=> `number`); the snapshot would have `number` where we need moment.

**Comment by mashaalmemon** [comment]
I found my issue. `observer` on `ItemRow` was necessary. Further, in my use case in a `useMemo` or `useEffect` if an individual property like `myVal` is used, it is not enough to have the dependency array contain only `item`; it should make a reference to individual properties we depend on like, which in the above case would be `item.myval`.

**Comment by mashaalmemon** [comment]
@xaviergonz Along the same vein is there a recommended way to have a model that is passed into a hook be observed as well? That is to say, we can use the `observer` hoc on components. Is there a corollary for hooks?

**Comment by xaviergonz** [comment]
As long as the hook (or rather, its list of dependencies) is inside an observer component it should be ok. 

E.g.:

```
// some Todo mobx-keystone model

const Component = observer(({todo}) => {
  useEffect(() => {
    // ...
  }, [ todo.text ]) // this should work ok as long as the component uses observer
})
```

**Comment by xaviergonz** [comment]
That being said, this won't work

```
const Component = observer(({todo}) => {
  useEffect(() => {
    // do something with todo.test
  }, [ todo])
})
```

since the todo ref itself won't change, and therefore won't trigger the effect

I recommend you to read this guide since it is common for mobx and mobx-keystone, etc

https://mobx.js.org/react-integration.html

**Comment by mashaalmemon** [comment]
Thanks for the examples. I've been burned by the observing the object vs the observable property problem before.

I've read the mobx link you sent extensively and learned quite a bit! However there are very few specific references to mobx inside custom hooks. The ones I've found are members of the community asking questions on stack overflow or the like:

* [https://github.com/mobxjs/mobx-react-lite/issues/67](https://github.com/mobxjs/mobx-react-lite/issues/67)
* [https://stackoverflow.com/questions/66302578/usage-of-mobx-in-custom-react-hooks](https://stackoverflow.com/questions/66302578/usage-of-mobx-in-custom-react-hooks)

I found from these resources that, as you mentioned, as long as the custom hook is used by a component that is an `observer` it will be aware of reactions.

It seems that `autorun` is the recommended way to have a `useEffect` react to observable changes vs what you suggested but it's a little foreign to developers who expect to have to specify dependency arrays!

Thanks again for your quick feedback. Much appreciated!
